crates/core/tedge_mapper/src/aws/converter.rs

```diff
@@ -44,13 +58,44 @@ impl Converter for AwsConverter {
     }
 
     async fn try_convert(&mut self, input: &Message) -> Result<Vec<Message>, Self::Error> {
-        let () = self.size_threshold.validate(input)?;
+        self.size_threshold.validate(input)?;
         let default_timestamp = self.add_timestamp.then(|| self.clock.now());
-        let mut serializer = ThinEdgeJsonSerializer::new_with_timestamp(default_timestamp);
-        let () = tedge_api::parser::parse_str(input.payload_str()?, &mut serializer)?;
 
-        let payload = serializer.into_string()?;
-        Ok(vec![(Message::new(&self.mapper_config.out_topic, payload))])
+        // serialize with ThinEdgeJson for measurements, for alarms and events just add the timestamp
+        let payload = if input.topic.name.starts_with("tedge/measurements") {
+            let mut serializer = ThinEdgeJsonSerializer::new_with_timestamp(default_timestamp);
+            tedge_api::parser::parse_str(input.payload_str()?, &mut serializer)?;
+
+            serializer.into_string()?
+        } else if input.topic.name.starts_with("tedge/events")
+            || input.topic.name.starts_with("tedge/alarms")
+            || input.topic.name.starts_with("tedge/health")
+        {
+            let mut payload_json: Map<String, Value> = serde_json::from_slice(&input.payload)?;
+
+            if let Some(timestamp) = default_timestamp {
+                let timestamp = timestamp
+                    .format(&time::format_description::well_known::Rfc3339)?
+                    .as_str()
+                    .into();
+                payload_json.entry("time").or_insert(timestamp);
+            }
+
+            serde_json::to_string(&payload_json)?
+        } else {
+            return Ok(vec![]);
+        };
+
+        let mut topic_iter = input.topic.name.split('/');
+        let td_type = topic_iter.nth(1).unwrap();
+        let mut topic_suffix = topic_iter.collect::<Vec<_>>().join("/");
```

Comment:
I feel suffix `/` is not needed here. Because the topic must be like `aws/td/measurement` right?

Reply:
The intent was to handle things like `tedge/alarms/<severity>/<alarm-type>/<child-id>`, etc. Then `td_type` would be `alarms`, and `topic_suffix` would be `/<severity>/<alarm-type>/<child-id>`, but you are correct that it seems to be needlessly complicated, I'm not sure why I did it like that, as just splitting on the first `/` character and taking the string to the right would work. Will fix.

Reply:
I now realize that by "suffix `/`" you meant that `.join("/")` at the end, which does not actually add text to the end of the string, but is actually [called on a slice](https://doc.rust-lang.org/stable/std/primitive.slice.html#method.join), to concatenate the elements with some string in between. So there is no `/` suffix there, but I've realized now that there is a better way to write it, so I'll change it

Reply:
@PradeepKiruvale should be fixed now

Reply:
But if you do split_once on `tedge/alarms/<severity>/<alarm-type>/<child-id>` then `td_type` will be `tedge` and `topic_suffix` will be `alarms/<severity>/<alarm-type>/<child-id>`. This does not address what you mentioned in your comment right?